### PR TITLE
feat: make lake run with non-root users

### LINKED ISSUE #5128: [Bug][External Plugins] External plugins' scripts cannot run as non-root
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-devlake/issues?q=is%3Aissue) and found no similar issues.

### What happened

Devlake cannot be run as non-root. When trying to run DevLake backend as non-root, it tries to load the external plugins by running some scripts (i.e. [azuredevops/run.sh](https://github.com/apache/incubator-devlake/blob/main/backend/python/plugins/azuredevops/run.sh) ) and fails with "Permission denied" message.

```
time="2023-05-08 09:53:17" level=info msg="Loading remote plugins"
time="2023-05-08 09:53:17" level=error msg="./run.sh: 20: poetry: Permission denied"
time="2023-05-08 09:53:17" level=error msg="failed to load plugins
	caused by: Error calling plugin-info
	Wraps: (2) get error when invoking remote function plugin-info
	Wraps: (3) remote error response:
	  | ./run.sh: 20: poetry: Permission denied
	Wraps: (4) exit status 127
	Wraps: (5) exit status 127
	Error types: (1) *hintdetail.withDetail (2) *hintdetail.withDetail (3) *hintdetail.withDetail (4) *hintdetail.withDetail (5) *exec.ExitError"
panic: Error calling plugin-info
Wraps: (2) get error when invoking remote function plugin-info
Wraps: (3) remote error response:
  | ./run.sh: 20: poetry: Permission denied
Wraps: (4) exit status 127
Wraps: (5) exit status 127
Error types: (1) *hintdetail.withDetail (2) *hintdetail.withDetail (3) *hintdetail.withDetail (4) *hintdetail.withDetail (5) *exec.ExitError

goroutine 1 [running]:
github.com/apache/incubator-devlake/server/services.Init()
	/app/server/services/init.go:90 +0x225
github.com/apache/incubator-devlake/server/api.CreateApiService()
	/app/server/api/api.go:61 +0x2e
main.main()
	/app/server/main.go:39 +0x7e
```

### What do you expect to happen

DevLake should be able to run as non-root.

### How to reproduce

Deploy Devlake in a kubernetes cluster (minikube) with helm adding these values:

```
lake:
  securityContext:
    fsGroup: 101
    runAsGroup: 1000
    runAsNonRoot: true
    runAsUser: 101
  
  containerSecurityContext: 
    allowPrivilegeEscalation: false
    capabilities:
      drop: 
          - all
```

You can also try to run it with Docker setting a non-root user.

### Anything else

I think it should be easy to fix by changing the ownership of the app folder in the Dockerfile. Maybe you could setup a specific non-root user to run the application.

### Version

main

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by ZhangNing10**
### ⚠️ Pre Checklist

> Please complete _ALL_ items in this checklist, and remove before submitting

- [ ] I have read through the [Contributing Documentation](https://devlake.apache.org/community/).
- [ ] I have added relevant tests.
- [ ] I have added relevant documentation.
- [ ] I will add labels to the PR, such as `pr-type/bug-fix`, `pr-type/feature-development`, etc.

### Summary
Make devlake run as non-root users

### Does this close any open issues?
Closes https://github.com/apache/incubator-devlake/issues/5128

### CODE REVIEW

**Comment by ZhangNing10** [comment]
@JorgeGar  @matrixji @keon94 @IronCore864 could you help review the pr? Thanks a ton!

**Comment by IronCore864** [comment]
@klesh please have a review. Run as non-root.

File: backend/Dockerfile [hunk 1]
```diff
@@ -132,19 +131,23 @@ RUN apt-get update && \
...
 # Setup Python Poetry package manager
 RUN curl -sSL https://install.python-poetry.org | python3 -
-RUN ln -sf /root/.local/bin/poetry /usr/local/bin
-# Build Python plugins
-RUN find /app/python/ -name "*.sh" | xargs -I{} chmod +x {}
-RUN /app/python/build.sh
+ENV PATH="$PATH:/home/devlake/.local/bin"
 
+# Build Python plugins, make sure the scripts has execute permission
+#RUN find /app/python/ -name "*.sh" | xargs -I{} chmod +x {}
```
Comment (IronCore864) [inline backend/Dockerfile:149]: missing space here.

**Review by IronCore864 — CHANGES_REQUESTED** [review]
Also, commit message lint failed.

File: backend/Dockerfile [hunk 2]
```diff
@@ -132,19 +131,23 @@ RUN apt-get update && \
 
 EXPOSE 8080
 
+RUN useradd -ms /bin/bash devlake -u 1010
+USER devlake
+ENV PYTHONUNBUFFERED=1
+
 WORKDIR /app
```
Comment (matrixji) [inline backend/Dockerfile:138]: Maybe we could set /app owned by devlake user.

**Review by matrixji — DISMISSED** [review]
LGTM

**Comment by klesh** [comment]
@keon94 @CamilleTeruel  Please take a look at this, would it affect the python plugin?

**Comment by keon94** [comment]
server starts fine but there's a log permissions issue:
image

**Comment by matrixji** [comment]
“Maybe we could set /app owned by devlake user.” => seems it's mandatory right now.

**Comment by matrixji** [comment]
> “Maybe we could set /app owned by devlake user.” => seems it's mandatory right now.
> 
> We could export /app/logs as a volume in Dockerfile we need a `RUN mkdir /app/logs` after `USER devlake` and use `VOLUME /app/logs` to export the /app/logs as a volume. After that:
> 
> * If the external volume is a named volume, then the owner of  /app/logs will keep devlake.
> * If the external volume is an anonymous volume (in our case, a local path), the owner of /app/logs will use the host folders, so we could use chown -R 1010:1010 on our dev host.
> 
> For the Kubernetes deployment, I think no other changes are needed.

@keon94 @ZhangNing10  could you also help verify it, I just have some simple test on local.

**Comment by ZhangNing10** [comment]
Thanks for the reviews, bros! @matrixji Thanks for your considerate suggestions, it looks very good! And I have adjusted accordingly, could you check again?

**Comment by klesh** [comment]
@keon94 Hi, can you help verify if it is working now?

**Comment by keon94** [comment]
@klesh No luck, same error. It looks like this [line](https://github.com/apache/incubator-devlake/blame/55e6835f87a752d1a182350fb0b1a6c34ff41c5f/backend/impls/logruslog/stream.go#L31) keeps creating a logs/ directory with 0755 permissions even though `os.ModePerm` is supposed to be an alias for 0777. 
image

**Comment by matrixji** [comment]
> * chown -R 1010:1010

**Comment by keon94** [comment]
@matrixji Yeah, turned out to be permission issue on my local logs directory. Just deleting it and letting docker-compose create a new one was sufficient.

**Review by keon94 — APPROVED** [review]

**Comment by awickham10** [comment]
As an FYI - I'm not sure this is totally working, unless it's not included in beta 1. Here's an excerpt from my logs.

```
Version:  v0.18.0-beta1@
time="2023-07-17 12:58:12" level=warning msg="Failed to create filestream for logs. Logs will not be piped to files.
	caused by: open /app/logs/devlake.log: permission denied
	Wraps: (2) open /app/logs/devlake.log
	Wraps: (3) permission denied
	Error types: (1) *hintdetail.withDetail (2) *fs.PathError (3) syscall.Errno"
time="2023-07-17 12:58:12" level=info msg="migration initialized"
time="2023-07-17 12:58:12" level=info msg="plugin loaded ae"
time="2023-07-17 12:58:12" level=info msg="plugin loaded bamboo"
time="2023-07-17 12:58:12" level=info msg="plugin loaded bitbucket"
time="2023-07-17 12:58:12" level=info msg="plugin loaded customize"
time="2023-07-17 12:58:13" level=info msg="plugin loaded dbt"
time="2023-07-17 12:58:13" level=info msg="plugin loaded dora"
time="2023-07-17 12:58:13" level=info msg="plugin loaded feishu"
time="2023-07-17 12:58:13" level=info msg="plugin loaded gitee"
time="2023-07-17 12:58:13" level=info msg="plugin loaded gitextractor"
time="2023-07-17 12:58:13" level=info msg="plugin loaded github"
time="2023-07-17 12:58:13" level=info msg="plugin loaded github_graphql"
time="2023-07-17 12:58:13" level=info msg="plugin loaded gitlab"
time="2023-07-17 12:58:13" level=info msg="plugin loaded icla"
time="2023-07-17 12:58:13" level=info msg="plugin loaded jenkins"
time="2023-07-17 12:58:13" level=info msg="plugin loaded jira"
time="2023-07-17 12:58:13" level=info msg="plugin loaded org"
time="2023-07-17 12:58:13" level=info msg="plugin loaded pagerduty"
time="2023-07-17 12:58:13" level=info msg="plugin loaded refdiff"
time="2023-07-17 12:58:13" level=info msg="plugin loaded slack"
time="2023-07-17 12:58:13" level=info msg="plugin loaded sonarqube"
time="2023-07-17 12:58:14" level=info msg="plugin loaded starrocks"
time="2023-07-17 12:58:14" level=info msg="plugin loaded tapd"
time="2023-07-17 12:58:14" level=info msg="plugin loaded teambition"
time="2023-07-17 12:58:14" level=info msg="plugin loaded trello"
time="2023-07-17 12:58:14" level=info msg="plugin loaded webhook"
time="2023-07-17 12:58:14" level=info msg="plugin loaded zentao"
time="2023-07-17 12:58:14" level=info msg="Loading remote plugins"
time="2023-07-17 12:58:14" level=error msg="./run.sh: 20: poetry: Permission denied"
time="2023-07-17 12:58:14" level=error msg="failed to load plugins
	caused by: Error calling plugin-info
	Wraps: (2) get error when invoking remote function plugin-info
	Wraps: (3) remote error response:
	  | ./run.sh: 20: poetry: Permission denied
	Wraps: (4) exit status 127
	Wraps: (5) exit status 127
	Error types: (1) *hintdetail.withDetail (2) *hintdetail.withDetail (3) *hintdetail.withDetail (4) *hintdetail.withDetail (5) *exec.ExitError"
panic: Error calling plugin-info
Wraps: (2) get error when invoking remote function plugin-info
Wraps: (3) remote error response:
  | ./run.sh: 20: poetry: Permission denied
Wraps: (4) exit status 127
Wraps: (5) exit status 127
Error types: (1) *hintdetail.withDetail (2) *hintdetail.withDetail (3) *hintdetail.withDetail (4) *hintdetail.withDetail (5) *exec.ExitError

goroutine 1 [running]:
github.com/apache/incubator-devlake/server/services.Init()
	/app/server/services/init.go:90 +0x24d
github.com/apache/incubator-devlake/server/api.CreateApiService()
	/app/server/api/api.go:61 +0x3d
main.main()
	/app/server/main.go:33 +0x36
```

**Comment by klesh** [comment]
@ZhangNing10 @IronCore864 please take a look, thanks.

**Comment by ZhangNing10** [comment]
@awickham10 devlake can only run with non-root user 1010, do you mean you need to run as another user?

**Comment by awickham10** [comment]
> hi @awickham10 devlake can only run with non-root user 1010, do you mean you need to run as another user?

I did manage to get it to run as 1010. Thank you!

**Comment by Feggah** [comment]
@ZhangNing10 How should I install DevLake with docker-compose to work around this error?

When I run docker-compose, the `logs` directory is created with `root` being the owner:
image

How can I specify in docker-compose to create the `logs` dir with `1010` as owner?

**Pull request merged**